### PR TITLE
Clear output of Torch SDPA for masked pieces

### PR DESCRIPTION
**Pull request by danieldk**
## Description

Since Torch 2.1, the Torch memory-efficient SDPA GPU kernel [returns NaN](https://github.com/pytorch/pytorch/issues/110213) for pieces that are completely masked out. This leads to NaN propagation in the next attention layer, because masked pieces get an attention of zero, but zero times NaN is still NaN.

In this we fix this by setting masked tokens to zero to clear out any NaNs.

We currently rely on the query dimension of the mask to be singular, but in the future we should probably redesign the `AttentionMask` class to account for the differences between attention masks and causal masks.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.

### CODE REVIEW

File: curated_transformers/layers/attention.py [hunk 1]
```diff
@@ -712,27 +715,59 @@ def forward(
...
+            #
+            # So we can only do this when we have attention masks where
+            # the query length it not specified, which are typically 'pure'
+            # attention masks (not causal maskes or combined masks):
+            #
+            # [batch_len, 1, 1, key_len]
+            #
+            # Doing this properly requires a redesign of our AttentionMask
+            # class.
+            assert attention_mask.bool_mask.size(-2) == 1
```
Comment (shadeMe) [inline curated_transformers/layers/attention.py:767]: We should add a message here as it's not an invariant at the moment.

```suggestion
            assert attention_mask.bool_mask.size(-2) == 1, "Torch SDP does not support attention masks with non-broadcastable query length yet"
```
